src/main/java/com/gregtechceu/gtceu/api/machine/feature/IRecipeLogicMachine.java

```diff
@@ -104,6 +104,50 @@ default void onWaiting() {
     default void afterWorking() {
         self().getDefinition().getAfterWorking().accept(this);
     }
+    @SuppressWarnings("unchecked")
+    default <T> T machineCallback(String str, @Nullable Object value, @Nullable T defaultValue) {
+        var callback = self().getDefinition().getCustomCallback().get(str);
+        if (callback != null) {
+            var res = callback.apply(this, value);
+            if(defaultValue==null)return (T) res;
+            if (res != null) {
+                if (defaultValue.getClass().isAssignableFrom(res.getClass())) {
+                    return (T) res;
+                } else {
+                    try {
+                        return (T) convertValue(res, defaultValue.getClass());
+                    } catch (Exception e) {
+                        GTCEu.LOGGER.error(e.getMessage());
+                        return defaultValue;
+                    }
+                }
+            }
+        }
+        return defaultValue;
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T convertValue(Object value, Class<T> targetClass) {
+        if (targetClass.isInstance(value)) {
+            return (T) value;
+        } else if (targetClass == int.class || targetClass == Integer.class) {
+            return (T) (Integer) Double.valueOf(value.toString()).intValue();
+        } else if (targetClass == long.class || targetClass == Long.class) {
+            return (T) (Long) Double.valueOf(value.toString()).longValue();
+        } else if (targetClass == float.class || targetClass == Float.class) {
+            return (T) (Float) Double.valueOf(value.toString()).floatValue();
+        } else if (targetClass == double.class || targetClass == Double.class) {
+            return (T) Double.valueOf(value.toString());
+        } else if (targetClass == boolean.class || targetClass == Boolean.class) {
+            return (T) Boolean.valueOf(value.toString());
+        } else if (targetClass == String.class) {
+            return (T) value.toString();
+        } else {
+            throw new IllegalArgumentException("Unsupported target class: " + targetClass);
+        }
+    }
+
```

Comment:
hell naw

Reply:
？

Reply:
that's really just... wrong. no. not merging that. no chance.

Reply:
Agreed